File: frontend/src/modules/proyectos/domain/MetaRepositoryPort.ts
import { Meta } from './Meta';
import { Componente } from './Componente';
import { Accion } from './Accion';

export interface MetaRepositoryPort {
  listar(proyectoId: string): Promise<Meta[]>;
  obtener(proyectoId: string, metaId: string): Promise<Meta>;
  crear(proyectoId: string, datos: { nombre: string; descripcion?: string }): Promise<Meta>;
  actualizar(proyectoId: string, metaId: string, datos: { nombre?: string; descripcion?: string }): Promise<Meta>;
  archivar(proyectoId: string, metaId: string): Promise<void>;
  eliminar(proyectoId: string, metaId: string): Promise<void>;
  listarComponentes(proyectoId: string, metaId: string): Promise<Componente[]>;
  crearComponente(proyectoId: string, metaId: string, datos: { nombre: string; descripcion?: string }): Promise<Componente>;
  listarAcciones(componenteId: string): Promise<Accion[]>;
  crearAccion(componenteId: string, datos: {
    nombre: string;
    descripcion?: string;
    unidadMedida?: string;
    proyeccion?: number;
    requisitosVerificacion?: import('./Accion').RequisitoVerificacion[];
    tiposEvidencia?: string[];
    startDate?: string | null;
    endDate?: string | null;
  }): Promise<Accion>;
  actualizarComponente(proyectoId: string, compId: string, datos: { nombre?: string; descripcion?: string }): Promise<Componente>;
  eliminarComponente(proyectoId: string, compId: string): Promise<void>;
  actualizarAccion(compId: string, accionId: string, datos: {
    nombre?: string;
    descripcion?: string;
    unidadMedida?: string;
    proyeccion?: number;
    tiposEvidencia?: string[];
    startDate?: string | null;
    endDate?: string | null;
  }): Promise<Accion>;
  eliminarAccion(compId: string, accionId: string): Promise<void>;
}

File: backend/modulos/metas/dominio/MetaRepositoryPort.py
from abc import ABC, abstractmethod
from typing import Optional, List
from .Entidades import Meta

class MetaRepositoryPort(ABC):
    @abstractmethod
    def crear(self, meta: Meta) -> Meta: pass
    @abstractmethod
    def obtener_por_id(self, id: str) -> Optional[Meta]: pass
    @abstractmethod
    def listar_por_proyecto(self, proyecto_id: str, incluir_archivadas: bool = False) -> List[Meta]: pass
    @abstractmethod
    def actualizar(self, meta: Meta) -> Meta: pass
    @abstractmethod
    def eliminar(self, id: str) -> bool: pass
    @abstractmethod
    def existen_componentes_por_meta(self, meta_id: str) -> bool: pass
